File: src/application/eval/evalService.ts
// ============================================================
// Eval Use Case（Application 层）
// 从 eval/route.ts 拆出的评估流程编排：
//   query 改写 → 实体关联文档（含 Wiki）/ 语义检索 → RAG 回答
//   → 输出 answer/contexts/sources/逐条分数链路
// ============================================================

import type { SearchResult } from '@/domain/search/types';
import type { ChunkStore } from '@/domain/document/types';
import type { StructQueryPort, EntityRepository } from '@/domain/entity/types';
import type { LlmClient, DocumentFileStore } from '../ports';
import type { HybridSearchFn } from '../search/hybridSearch';
import type { SmartRewriter } from '../search/queryRewriter';
import type { RagChatStreamFn } from '../chat/ragEngine';
import { loadEntityDocsContentForEval, filterChunksByDocTypes } from '../chat/entityDocs';

export interface EvalRequest {
  query: string;
  topK?: number;
  llm?: LlmClient;
}

export interface EvalResult {
  query: string;
  answer: string;
  contexts: string[];
  sources: string[];
  searchMethod: 'rrf' | 'entity';
  numResults: number;
  matchedEntities: string[];
  /** 逐条分数链路（score/scores 与 contexts 一一对应），供评估追溯 RRF→rerank 打分过程 */
  resultScores: Array<{ score: number; scores: SearchResult['scores'] }>;
}

/** 客户企业特征正则（判断是否走结构化检索） */
const ENTERPRISE_PATTERN = /集团|公司|银行|证券|电力|保险|能源|通信|钢铁|船舶|置地|宝武|中车|中化|中钢|招商局|华润|中信|浦发|招商|万科|中粮|南方电网|国家电网|中国移动|中国联通|中国电信|中国银行|建设银行|农业银行|工商银行|交通银行|国泰君安|华泰|光大|民生|平安|太平洋|新华|人寿/;

export interface EvalService {
  evaluate(req: EvalRequest): Promise<EvalResult>;
}

export function createEvalService(deps: {
  llm: LlmClient;
  chunkStore: ChunkStore;
  structQuery: StructQueryPort;
  entityRepo: EntityRepository;
  fileStore: DocumentFileStore;
  hybridSearch: HybridSearchFn;
  smartRewriter: SmartRewriter;
  ragChatStream: RagChatStreamFn;
}): EvalService {
  const {
    llm: defaultLlm,
    chunkStore,
    structQuery,
    entityRepo,
    fileStore,
    hybridSearch,
    smartRewriter,
    ragChatStream,
  } = deps;

  function extractEnterpriseEntities(matchedEntities: string[]): string[] {
    if (!structQuery.isReady()) return [];

    const knownEntities = entityRepo.getKnownEntities();
    const entityMap = new Map<string, { type: string; category: string }>();
    for (const e of knownEntities) {
      entityMap.set(e.name.toLowerCase(), { type: e.type, category: e.category });
    }

    const results: string[] = [];
    for (const entity of matchedEntities) {
      const info = entityMap.get(entity.toLowerCase());
      if (info && info.type === 'entity' && ENTERPRISE_PATTERN.test(entity)) {
        results.push(entity);
      }
    }

    return results;
  }

  function shouldUseStructuredSearch(matchedEntities: string[]): boolean {
    return extractEnterpriseEntities(matchedEntities).length > 0;
  }

  return {
    async evaluate(req: EvalRequest): Promise<EvalResult> {
      const {
        query,
        topK = 10,
        llm,
      } = req;
      const clientLlm = llm ?? defaultLlm;

      const trimmedQuery = query.trim();

      const rewriteResult = await smartRewriter.rewrite(trimmedQuery, {
        llm: clientLlm,
        previousQuery: undefined,
      });
      const matched = rewriteResult.entities;
      const rewrittenQuery = rewriteResult.rewrittenQuery;

      let results: SearchResult[] = [];
      let entityDocsContent: string | undefined;
      let entitySources: string[] = [];
      let searchMethod: 'rrf' | 'entity' = 'rrf';

      // 实体关联文档加载（完整策略：Wiki 词条 + 短文档全文，长文档片段提取）
      // 使用所有匹配的实体进行 OR 查询，确保精确匹配到目标文档
      if (matched.length > 0 && shouldUseStructuredSearch(matched)) {
        const entityResult = await loadEntityDocsContentForEval(structQuery, fileStore, matched);
        if (entityResult) {
          entityDocsContent = entityResult.docsContent;
          entitySources = entityResult.sources;
          searchMethod = 'entity';
        }
      }

      // 降级为语义检索
      if (!entityDocsContent) {
        const filteredChunkIds = rewriteResult.relevantDocTypes?.length > 0
          ? filterChunksByDocTypes(chunkStore, rewriteResult.relevantDocTypes, '[Eval]')
          : null;
        const semanticResults = await hybridSearch(rewrittenQuery || trimmedQuery, topK, 20, 20, {
          matchedKeywords: matched.length > 0 ? matched : undefined,
          filteredChunkIds: filteredChunkIds ?? undefined,
        });
        results = semanticResults;
        searchMethod = 'rrf';
      }

      let answer = '';
      let finalResults: SearchResult[] = results;

      if (results.length > 0 || entityDocsContent) {
        const generator = ragChatStream(trimmedQuery, {
          llm: clientLlm,
          topK,
          entityDocsContent,
          preSearchResults: results.length > 0 ? results : undefined,
        });

        let fullAnswer = '';
        for await (const chunk of generator) {
          if (chunk.type === 'token' && chunk.content) {
            fullAnswer += chunk.content;
          } else if (chunk.type === 'context' && chunk.results) {
            finalResults = chunk.results;
          }
        }
        answer = fullAnswer || '未能生成回答';
      } else {
        answer = '未检索到相关资料，请尝试其他关键词。';
      }

      const contextChunks = finalResults.slice(0, topK).map(r => r.chunk?.content || '').filter(Boolean);
      const contextSources = finalResults.slice(0, topK).map(r => r.chunk?.docTitle || 'unknown').filter(Boolean);

      // 如果有实体文档内容，作为主要上下文
      if (entityDocsContent) {
        contextChunks.unshift(entityDocsContent.slice(0, 5000)); // 扩大上下文长度到 5000 字符
      }

      // 根据检索类型确定结果数量和来源
      let finalNumResults = finalResults.length;
      let finalSources = contextSources;
      if (searchMethod === 'entity' && entitySources.length > 0) {
        finalNumResults = entitySources.length;
        finalSources = entitySources;
      }

      return {
        query: trimmedQuery,
        answer,
        contexts: contextChunks,
        sources: finalSources,
        searchMethod,
        numResults: finalNumResults,
        matchedEntities: matched,
        resultScores: finalResults.slice(0, topK).map(r => ({
          score: r.score,
          scores: r.scores ?? {},
        })),
      };
    },
  };
}

File: server/rag/eval/eval_service.py
"""Eval Use Case（逐字移植自 src/application/eval/evalService.ts）。

query 改写 → 企业实体门禁 → 实体关联文档（eval 变体，含 Wiki）/ 语义检索
→ RAG 回答收集 → answer/contexts/sources/逐条分数链路。

与 chat 的差异：无 session、无缓存、无追问、无对话历史；
ragChatStream 只传 {llm, topK, rerankTopK, entityDocsContent, preSearchResults}。
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from ..chat.entity_docs import (
    filter_chunks_by_doc_types,
    load_entity_docs_content_for_eval,
)
from ..chat.query_rewriter import SmartRewriteOptions
from ..chat.rag_engine import RagChatOptions
from ..chat.types import RagContextEvent, RagTokenEvent
from ..retrieve.hybrid_search import HybridSearchOptions
from ..retrieve.profile import resolve_profile
from ..types import scores_to_dict

# 客户企业特征正则（判断是否走结构化检索，逐字搬运，勿"补全"）
ENTERPRISE_PATTERN = re.compile(
    r"集团|公司|银行|证券|电力|保险|能源|通信|钢铁|船舶|置地|宝武|中车|中化|中钢|"
    r"招商局|华润|中信|浦发|招商|万科|中粮|南方电网|国家电网|中国移动|中国联通|"
    r"中国电信|中国银行|建设银行|农业银行|工商银行|交通银行|国泰君安|华泰|光大|"
    r"民生|平安|太平洋|新华|人寿"
)


@dataclass
class EvalRequestOptions:
    query: str
    top_k: int = 10
    llm: object | None = None
    profile_id: str | None = None


@dataclass
class EvalResult:
    query: str
    answer: str
    contexts: list[str]
    sources: list[str]
    search_method: str  # 'rrf' | 'entity'
    num_results: int
    matched_entities: list[str] = field(default_factory=list)
    profile_id: str = "baseline"
    result_scores: list[dict] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "query": self.query,
            "answer": self.answer,
            "contexts": self.contexts,
            "sources": self.sources,
            "searchMethod": self.search_method,
            "numResults": self.num_results,
            "matchedEntities": self.matched_entities,
            "profileId": self.profile_id,
            "resultScores": self.result_scores,
        }


class EvalService:
    def __init__(self, deps: dict) -> None:
        self._default_llm = deps["llm"]
        self._chunk_store = deps["chunk_store"]
        self._struct_query = deps["struct_query"]
        self._entity_repo = deps["entity_repo"]
        self._file_store = deps["file_store"]
        self._hybrid_search = deps["hybrid_search"]
        self._smart_rewriter = deps["smart_rewriter"]
        self._rag_chat_stream = deps["rag_chat_stream"]

    def _extract_enterprise_entities(self, matched_entities: list[str]) -> list[str]:
        if not self._struct_query.is_ready():
            return []

        entity_map = {
            e["name"].lower(): {"type": e.get("type"), "category": e.get("category", "")}
            for e in self._entity_repo.get_known_entities()
        }

        results: list[str] = []
        for entity in matched_entities:
            info = entity_map.get(entity.lower())
            if (
                info
                and info["type"] == "entity"
                and ENTERPRISE_PATTERN.search(entity)
            ):
                results.append(entity)
        return results

    def _should_use_structured_search(self, matched_entities: list[str]) -> bool:
        return len(self._extract_enterprise_entities(matched_entities)) > 0

    def evaluate(self, req: EvalRequestOptions) -> EvalResult:
        client_llm = req.llm or self._default_llm
        top_k = req.top_k or 10
        profile = resolve_profile(req.profile_id)
        trimmed_query = req.query.strip()

        rewrite_result = self._smart_rewriter.rewrite(
            trimmed_query,
            SmartRewriteOptions(llm=client_llm, previous_query=None),
        )
        matched = rewrite_result.entities
        rewritten_query = rewrite_result.rewritten_query

        results = []
        entity_docs_content: str | None = None
        entity_sources: list[str] = []
        search_method = "rrf"

        # 实体关联文档（eval 完整策略：Wiki 词条 + 短文档全文/长文档片段）
        if matched and self._should_use_structured_search(matched):
            entity_result = load_entity_docs_content_for_eval(
                self._struct_query, self._file_store, matched
            )
            if entity_result:
                entity_docs_content = entity_result.docs_content
                entity_sources = entity_result.sources
                search_method = "entity"

        # 降级语义检索
        if not entity_docs_content:
            if rewrite_result.relevant_doc_types:
                filtered_chunk_ids = filter_chunks_by_doc_types(
                    self._chunk_store, rewrite_result.relevant_doc_types, "[Eval]"
                )
            else:
                filtered_chunk_ids = None
            results = self._hybrid_search(
                rewritten_query or trimmed_query,
                top_k,
                20,
                20,
                HybridSearchOptions(
                    matched_keywords=matched if matched else None,
                    filtered_chunk_ids=filtered_chunk_ids,
                    profile=profile,
                ),
            )
            search_method = "rrf"

        answer = ""
        final_results = results

        if results or entity_docs_content:
            generator = self._rag_chat_stream(
                trimmed_query,
                RagChatOptions(
                    llm=client_llm,
                    top_k=top_k,
                    rerank_top_k=profile.rerank_top_k,
                    entity_docs_content=entity_docs_content,
                    pre_search_results=results if results else None,
                ),
            )

            full_answer = ""
            for event in generator:
                if isinstance(event, RagTokenEvent) and event.content:
                    full_answer += event.content
                elif isinstance(event, RagContextEvent):
                    final_results = event.results
            answer = full_answer or "未能生成回答"
        else:
            answer = "未检索到相关资料，请尝试其他关键词。"

        context_chunks = [
            r.chunk.content
            for r in final_results[:top_k]
            if r.chunk and r.chunk.content
        ]
        context_sources = [
            r.chunk.doc_title or "unknown" if r.chunk else "unknown"
            for r in final_results[:top_k]
        ]
        context_sources = [s for s in context_sources if s]

        # 实体文档内容作为首要上下文（先 slice 再 unshift，截断 5000 字符）
        if entity_docs_content:
            context_chunks = [entity_docs_content[:5000]] + context_chunks

        final_num_results = len(final_results)
        final_sources = context_sources
        if search_method == "entity" and entity_sources:
            final_num_results = len(entity_sources)
            final_sources = entity_sources

        return EvalResult(
            query=trimmed_query,
            answer=answer,
            contexts=context_chunks,
            sources=final_sources,
            search_method=search_method,
            num_results=final_num_results,
            matched_entities=matched,
            profile_id=profile.id,
            result_scores=[
                {"score": r.score, "scores": scores_to_dict(r.scores)}
                for r in final_results[:top_k]
            ],
        )


def create_eval_service(deps: dict) -> EvalService:
    return EvalService(deps)
